prow/metrics/metrics_test.go

```diff
@@ -0,0 +1,193 @@
+		}
+	}
+}
+
+func TestMetricsPusherEndpoints(t *testing.T) {
+	endpoints := []string{"ep0", "ep1", "ep9000"}
+	actualEndpoints := make(chan string)
+
+	endpointCheckGatherer := func(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+		actualEndpoints <- url
+		return nil
+	}
+
+	pusher := Pusher{
+		gatherer: endpointCheckGatherer,
+		configAgent: &fakeConfigAgent{
+			endpoints: endpoints,
+		},
+	}
+	go pusher.Start("")
+
+	for _, expectedEndpoint := range endpoints {
+		actualEndpoint := <-actualEndpoints
+		if actualEndpoint != expectedEndpoint {
+			t.Errorf("Expected endpoint to be '%s', got '%s'", expectedEndpoint, actualEndpoint)
+		}
+	}
+}
+
+func TestMetricsPusherWaiterCalls(t *testing.T) {
+	intervals := []time.Duration{1 * time.Second, 2 * time.Second}
+
+	fakeGatherer := func(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+		return nil
+	}
+
+	ranFuncCh := make(chan time.Duration)
+	waiter := func(d time.Duration) <-chan time.Time {
+		waiterCh := make(chan time.Time, 1)
+		ranFuncCh <- d
+		waiterCh <- time.Time{}
+		return waiterCh
+	}
+
+	pusher := Pusher{
+		gatherer: fakeGatherer,
+		configAgent: &fakeConfigAgent{
+			intervals: intervals,
+		},
+		waiter: waiter,
+	}
+
+	go pusher.Start("")
+
+	for _, expectedInterval := range intervals {
+		actualInterval := <-ranFuncCh
+		if actualInterval != expectedInterval {
+			t.Errorf("Expected interval was '%s', got '%s'", expectedInterval, actualInterval)
+		}
+	}
+}
+
+func TestMetricsPusherEmptyEndpoint(t *testing.T) {
+	endpoints := []string{"", "ep0", "", "ep1"}
+
+	ranFuncCh := make(chan string)
+	fakeGatherer := func(job string, grouping map[string]string, url string, g prometheus.Gatherer) error {
+		ranFuncCh <- url
+		return nil
+	}
+
+	pusher := Pusher{
+		gatherer: fakeGatherer,
+		configAgent: &fakeConfigAgent{
+			endpoints: endpoints,
+		},
+	}
+
+	go pusher.Start("")
+
+	collectedEndpoints := []string{}
+	collectedEndpoints = append(collectedEndpoints, <-ranFuncCh)
+	collectedEndpoints = append(collectedEndpoints, <-ranFuncCh)
+
+	expectedEndpoints := []string{"ep0", "ep1"}
+
+	if !reflect.DeepEqual(collectedEndpoints, expectedEndpoints) {
+		t.Errorf("Expected collected endpoints to be '%v', got '%v'", expectedEndpoints, collectedEndpoints)
+	}
+}
+
+type fakeConfigAgent struct {
+	endpoints         []string
+	intervals         []time.Duration
+	calledForEndpoint bool
+	endpointIdx       int
+	intervalIdx       int
+}
+
+// Config returns a minimal *config.Config, filled only with the bits
+// interesting for the metrics pusher.
+// We also take advantage of the fact that Config() is *always* called twice
+// per pusher iteration:
+//   1st call to get the currently configured interval
+//   2nd call to get the currently configured endpoint
+// Therefore we will also return different configs per call. On every even call
```

Comment:
This seems like a lot of baked-in logic for a test harness. Having a simpler interface may make the tests easier to understand and work with.